File: src/leetcode_grind/day_611.rs
// https://leetcode.com/problems/find-valid-matrix-given-row-and-column-sums/description/?envType=daily-question&envId=2024-07-20
pub fn restore_matrix_i(row_sum: Vec<i32>, col_sum: Vec<i32>) -> Vec<Vec<i32>> {
    let n = row_sum.len();
    let m = col_sum.len();

    let mut curr_row_sum = vec![0; n];
    let mut curr_col_sum = vec![0; m];

    let mut orig_matrix = vec![vec![0; m]; n];
    for i in 0..n {
        for j in 0..m {
            orig_matrix[i][j] = (row_sum[i] - curr_row_sum[i]).min(col_sum[j] - curr_col_sum[j]);
            curr_row_sum[i] += orig_matrix[i][j];
            curr_col_sum[j] += orig_matrix[i][j];
        }
    }
    orig_matrix
}

pub fn restore_matrix_ii(mut row_sum: Vec<i32>, mut col_sum: Vec<i32>) -> Vec<Vec<i32>> {
    let n = row_sum.len();
    let m = col_sum.len();

    let mut orig_matrix = vec![vec![0; m]; n];
    let mut i = 0;
    let mut j = 0;

    while i < n && j < m {
        orig_matrix[i][j] = row_sum[i].min(col_sum[j]);

        row_sum[i] -= orig_matrix[i][j];
        col_sum[j] -= orig_matrix[i][j];

        if row_sum[i] == 0 {
            i += 1;
        } else {
            j += 1;
        }
    }

    orig_matrix
}

File: java/leetcode_grind/Day611.java
package leetcode_grind;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Day611 {

    // https://leetcode.com/problems/build-a-matrix-with-conditions/description/?envType=daily-question&envId=2024-07-21
    static class Solution1 {
        public int[][] buildMatrix(int k, int[][] rowConditions, int[][] colConditions) {
            List<Integer> orderRows = topoSort(rowConditions, k);
            List<Integer> orderColumns = topoSort(colConditions, k);

            if (orderRows.isEmpty() || orderColumns.isEmpty())
                return new int[0][0];

            int[][] matrix = new int[k][k];
            for (int i = 0; i < k; i++) {
                for (int j = 0; j < k; j++) {
                    if (orderRows.get(i).equals(orderColumns.get(j))) {
                        matrix[i][j] = orderRows.get(i);
                    }
                }
            }
            return matrix;
        }

        private List<Integer> topoSort(int[][] edges, int n) {
            List<List<Integer>> adj = new ArrayList<>();
            for (int i = 0; i <= n; i++) {
                adj.add(new ArrayList<>());
            }
            for (int[] edge : edges) {
                adj.get(edge[0]).add(edge[1]);
            }

            List<Integer> order = new ArrayList<>();
            int[] visited = new int[n + 1];
            boolean[] hasCycle = { false };

            for (int i = 1; i <= n; i++) {
                if (visited[i] == 0) {
                    dfs(i, adj, visited, order, hasCycle);
                    if (hasCycle[0])
                        return new ArrayList<>();
                }
            }

            Collections.reverse(order);
            return order;
        }

        private void dfs(int node, List<List<Integer>> adj, int[] visited, List<Integer> order, boolean[] hasCycle) {
            visited[node] = 1;
            for (int neighbor : adj.get(node)) {
                if (visited[neighbor] == 0) {
                    dfs(neighbor, adj, visited, order, hasCycle);
                    if (hasCycle[0])
                        return;
                } else if (visited[neighbor] == 1) {
                    hasCycle[0] = true;
                    return;
                }
            }
            visited[node] = 2;
            order.add(node);
        }
    }
}
